aten/src/ATen/native/cuda/ForeachUnaryOp.cu

```diff
@@ -180,150 +204,164 @@ void foreach_unary_op_bfloat16_(TensorList tensors) {
-    }                                                                   \
-                                                                        \
-    return foreach_unary_op_bfloat16<NAME1>(tensors);                   \
-}                                                                       \
-                                                                        \
-void foreach_tensor_##NAME##_cuda_(TensorList tensors) {                \
-    check_foreach_api_restrictions(tensors);                            \
-                                                                        \
-    if (!can_use_fast_route(tensors)) {                                 \
-        return at::native::foreach_tensor_##NAME##_slow_(tensors);      \
-    }                                                                   \
-                                                                        \
-    foreach_unary_op_bfloat16_<NAME1>(tensors);                         \
+#define FLOATING_HALF(NAME, NAME1)                                                 \
+template<typename T>                                                               \
+struct NAME1 {                                                                     \
+    __device__ T operator()(T t) const { return std::NAME(t); }                    \
+};                                                                                 \
+                                                                                   \
+std::vector<Tensor> foreach_tensor_##NAME##_cuda(TensorList tensors) {             \
+    check_foreach_api_restrictions(tensors);                                       \
+    if (!can_use_fast_route(tensors)) {                                            \
+        return at::native::foreach_tensor_##NAME##_slow(tensors);                  \
+    }                                                                              \
+                                                                                   \
+                                                                                   \
+    return floating_half<NAME1>(tensors);                                          \
+}                                                                                  \
+                                                                                   \
+void foreach_tensor_##NAME##_cuda_(TensorList tensors) {                           \
+    check_foreach_api_restrictions(tensors);                                       \
+    if (!can_use_fast_route(tensors)) {                                            \
+        at::native::foreach_tensor_##NAME##_slow_(tensors);                        \
+    }                                                                              \
+                                                                                   \
+    floating_half_<NAME1>(tensors);                                                \
```

Comment:
It seems to me what you really want here is three separate macros:
```
// makes the functor
#define STD_FUNCTOR(op_name, functor_name)
template<typename T>
struct NAME1 {
    __device__ T operator()(T t) const { return std::NAME(t); }
};

// given a functor and a "dispatch function", creates the outplace and inplace operations
#define OP_CUSTOM_FUNCTOR(function, op_name, functor_name)
std::vector<Tensor> foreach_tensor_##op_name##_cuda(TensorList tensors) {
    check_foreach_api_restrictions(tensors);
    if (!can_use_fast_route(tensors)) {
        return at::native::foreach_tensor_##op_name##_slow(tensors);
    }
    return function<functor_name>(tensors);
}
void foreach_tensor_##op_name##_cuda_(TensorList tensors) {
    check_foreach_api_restrictions(tensors);
    if (!can_use_fast_route(tensors)) {
        return at::native::foreach_tensor_##op_name##_slow_(tensors);
    }

    ##function##_<functor_name>(tensors);
}

// creates a functor, outplace version, and inplace version.
#define OP(function, op_name, functor_name)
STD_FUNCTOR(op_name, functor_name);
OP_CUSTOM_FUNCTOR(function, op_name, functor_name);
```

And then you can write code like:
```
OP(floating_half, erfc, Erfc);
OP(floating_half, expm1, Expm1);
OP(floating_half_bfloat16, log1p, Log1p);
...
```